core/sync/chan/chan.odin

```diff
@@ -633,6 +634,11 @@ send_raw :: proc "contextless" (c: ^Raw_Chan, msg_in: rawptr) -> (ok: bool) {
 			sync.signal(&c.r_cond)
 		}
 		sync.wait(&c.w_cond, &c.mutex)
+
+		if c.closed {
+			return false
```

Comment:
Could this now return false if it sent the message and it was also closed?

Reply:
Yes good point, that does seem possible.

If you have three threads: the first calls `send` and the other two call `recv` and `close`, it's possible that `recv` executes (copies the value) and then `close` grabs the lock before `send` can wakeup. In that case the value was actually received but the sender reports false because the close call sneaked in between the recv and the send wakeup.

Good: `send: sleep` -> `recv` -> `send: wake` -> `close`
Bad:  `send: sleep` -> `recv` -> `close` -> `send: wake`

Reply:
One possible solve is to introduce some state, like a boolean `did_recv`. The `recv` proc sets the `did_recv` to true after doing the copy, but before signalling the sender. If the sender wakes up and `did_recv` is true then it can know that it was sent, even if the channel was also closed.

Reply:
Or we can make the `send` proc wait for a reader in the same way that `recv` waits for a writer.

Reply:
```go
		for !c.closed &&
			c.w_waiting == 0 {
			c.r_waiting += 1
			sync.wait(&c.r_cond, &c.mutex)
			c.r_waiting -= 1
		}
```